servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -154,188 +165,288 @@
             final int linearSearchSpace,
             @Nullable final HealthCheckConfig healthCheckConfig) {
         this.targetResource = requireNonNull(targetResourceName) + " (instance @" + toHexString(hashCode()) + ')';
-        Processor<Object, Object> eventStreamProcessor = newPublisherProcessorDropHeadOnOverflow(32);
+        this.eventPublisher = requireNonNull(eventPublisher);
         this.eventStream = fromSource(eventStreamProcessor);
         this.connectionFactory = requireNonNull(connectionFactory);
         this.linearSearchSpace = linearSearchSpace;
+        this.healthCheckConfig = healthCheckConfig;
+        this.asyncCloseable = toAsyncCloseable(graceful -> {
+            discoveryCancellable.cancel();
+            eventStreamProcessor.onComplete();
+            final CompositeCloseable compositeCloseable;
+            for (;;) {
+                List<Host<ResolvedAddress, C>> currentList = usedHosts;
+                if (isClosedList(currentList) ||
+                        usedHostsUpdater.compareAndSet(this, currentList, new ClosedList<>(currentList))) {
+                    compositeCloseable = newCompositeCloseable().appendAll(currentList).appendAll(connectionFactory);
+                    LOGGER.debug("Load balancer for {} is closing {}gracefully. Last seen addresses (size={}): {}.",
+                            targetResource, graceful ? "" : "non", currentList.size(), currentList);
+                    break;
+                }
+            }
+            return (graceful ? compositeCloseable.closeAsyncGracefully() : compositeCloseable.closeAsync())
+                    .beforeOnError(t -> {
+                        if (!graceful) {
+                            usedHosts = new ClosedList<>(emptyList());
+                        }
+                    })
+                    .beforeOnComplete(() -> usedHosts = new ClosedList<>(emptyList()));
+        });
+        subscribeToEvents(false);
+    }
 
-        toSource(eventPublisher).subscribe(
-                new Subscriber<Collection<? extends ServiceDiscovererEvent<ResolvedAddress>>>() {
+    private void subscribeToEvents(boolean resubscribe) {
```

Comment:
is there a possibility that this method is called multiple times concurrently (and if so, do you think it has racy conditions around cancelation?)

Reply:
`nextResubscribeTime` is used to make sure there is only one thread at a time that performs re-subscribe (unless I messed up something around lines 540-543). I will add `assert nextResubscribeTime == RESUBSCRIBING` at the beginning of this method.